### PR TITLE
Add Optimizer Class & separate math functions

### PR DESCRIPTION
**Pull request by jijoongmoon**
- Add Optimizer Class and remove duplicated codes.
  . Hyperparameters related with Optimizer class placed Optimizer
  Class.
    . OptParam, WeightDecayParam, OptType, WeightDecayType

- Seperate math functions in util_func.h/cpp

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: nntrainer/src/util_func.cpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+#include "util_func.h"
+#include <assert.h>
+#include "math.h"
+#include "tensor.h"
+
+Tensors::Tensor softmaxPrime(Tensors::Tensor x) {
+  int batch = x.getBatch();
+  int width = x.getWidth();
+  int height = x.getHeight();
+  assert(height == 1);
```
Comment (dongju-chae) [inline nntrainer/src/util_func.cpp:34]: Minor comment: If the height is always 1, I think you don't need to keep `height` in the below loops (for readability).

Comment (jijoongmoon) [reply nntrainer/src/util_func.cpp:34]: Yes. But this is just temporary. Once TensorDim Class is implemented successfully, then this alert will be deleted.

Comment (dongju-chae) [reply nntrainer/src/util_func.cpp:34]: Got it

File: nntrainer/src/neuralnet.cpp [hunk 2]
```diff
@@ -93,47 +93,44 @@ std::vector<std::string> parseLayerName(std::string ll) {
...
   int ret;
   unsigned int i;
-
   /**
    * @brief     Optimizer String from configure file
    *            "sgd"  : Stochestic Gradient Descent
    *            "adam" : Adaptive Moment Estimation
    */
-  std::array<std::string, 150> Optimizer_string = {"sgd", "adam", "unknown"};
+  std::array<std::string, 150> Optimizer_string = {"sgd", "adam", "unkown"};
```
Comment (dongju-chae) [inline nntrainer/src/neuralnet.cpp:101]: Rollback to "unkown"??

Comment (jijoongmoon) [reply nntrainer/src/neuralnet.cpp]: Ooop. Fixed. Thanks.

File: nntrainer/include/layers.h [hunk 3]
```diff
@@ -177,7 +140,7 @@ class Layer {
    * @brief     Optimizer Setter
    * @param[in] opt Optimizer
    */
-  virtual void setOptimizer(Optimizer opt) = 0;
+  void setOptimizer(Optimizer opt);
```
Comment (myungjoo) [inline nntrainer/include/layers.h:143]: Add virtual?

What do you want?
```
Layer * a = new InputLayer();
a->setOptimizer();
```
Do you want to access ```Layer::setOptimizer()``` (current code)? or ```InputLayer::setOptimizer()``` (add virtual)?

File: nntrainer/src/neuralnet.cpp [hunk 4]
```diff
@@ -121,7 +120,6 @@ unsigned int parseType(std::string ll, input_type t) {
    *            "tanh"  : tanh
    *            "sigmoid" : sigmoid
    *            "relu" : relu
-   *            "softmax" : softmax
```
Comment (helloahn) [inline nntrainer/src/neuralnet.cpp:124]: Is there any reason removing softmax from the comment? `activation_string` still has softmax.

Comment (jijoongmoon) [reply nntrainer/src/neuralnet.cpp]: Yes. fixed. thanks.

File: nntrainer/include/layers.h (hunk 3 above)
Comment (jijoongmoon) [reply nntrainer/include/layers.h]: virtual added. Thanks.

File: nntrainer/include/optimizer.h [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+ */
+typedef struct {
+  float learning_rate;
+  double beta1;
+  double beta2;
+  double epsilon;
+  float decay_rate;
+  float decay_steps;
+  weight_decay_param weight_decay;
+} OptParam;
```
Comment (wooksong) [inline nntrainer/include/optimizer.h:73]: Q: "weight_decay_param" is follwed by the snake naming convention while OptPram is followed by the camel naming convention. Is there any difference between two typedefed structures?

Comment (jijoongmoon) [reply nntrainer/include/optimizer.h:73]: I will change it later for all codes. But the rule is as below.
- Type Names : Pascal Naming rule ( OptParam )
- Variable Names : snake Naming rule ( weight_decay )
- Function Name : Camel Naming rule (setFunction())
- Enumerator : Upper Capital with underscore.
if you have better suggestion, then please let me know.

File: nntrainer/include/optimizer.h [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+ * @brief     Enumeration of Optimizer
+ *            0. SGD
+ *            1. ADAM
+ *            2. Unknown
+ */
+enum class OptType
+{
+  sgd = 0,
+  adam = 1,
+  unkonwn = 2
```
Comment (myungjoo) [inline nntrainer/include/optimizer.h:38]: ```s/unkonwn/unknown```

File: nntrainer/include/optimizer.h [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+ * @author	Jijoong Moon <jijoong.moon@samsung.com>
+ * @bug		No known bugs except for NYI items
+ *
+ */
+#ifndef __OPTIMIZER_H__
+#define __OPTIMIZER_H__
+
+#include <iostream>
+#include "tensor.h"
+
```
Comment (myungjoo) [inline nntrainer/include/optimizer.h:27]: Please consider applying namespace.

Comment (myungjoo) [reply nntrainer/include/optimizer.h:27]: And please prevent C code including this C++-only header. (use .hh as extension, add __cplusplus macro)

Comment (jijoongmoon) [reply nntrainer/include/optimizer.h:27]: I will apply them later. Thanks.
